### ISSUE TITLE
Get dependency information from idea libs?

### ISSUE THREAD

**Issue by loosebazooka** [JetBrains/intellij-platform-gradle-plugin#68]
Presumably this would a large amount of work but, It'd be nice if we could automatically detect conflicts on libraries included with the idea SDK.

For instance someone including `'org.apache.httcomponents:httpclient:4.01'` has a conflict with `idea-xx-123.123.23/lib/httpclient-4.4.1.jar`. It would be nice if `configurations.resolutionStrategy.failOnVersionConflict()` has more info about these dependencies to have an explicit build failure and potentially prevent a runtime conflict.

Right now we're doing something kind of like this : 

```
configuration {
  resolutionStrategy {
        failOnVersionConflict()

        def provided = [["org.apache.httpcomponents", "httpclient"],
                        ["com.google.guava", "guava"]]
        eachDependency { DependencyResolveDetails dependency ->
            if (provided.contains([dependency.requested.group, dependency.requested.name])) {
                throw new GradleException("build failure")
            }
        }
    }
}
```

But it seems like a lot of work to really cover everything in idea/lib?

**Comment by zolotov** [comment]
Yes, first of all it's quite a lot of work to cover everything in idea/lib and I don't think that it should be done in this plugin, at least in its current state. When IntelliJ will be migrated to the maven/gradle model (and I hope it will happen eventually) we'll get conflicts resolving automatically.

In the second, have you ever got runtime conflicts in such cases?

**Comment by loosebazooka** [comment]
I can trigger the build failure if I include those libraries. I would have to do something insane to verify all the libraries.  The only thing I can think of is parsing filenames in the idea/lib folder and trying to correctly guess the dependency information (that's for another day)

**Comment by zolotov** [comment]
I see. 

At the moment it's not even possible to automatically retrieve `group-id` for every jar in `idea/lib/`. So if someone ready to make some mapping like `filename.jar -> group:artifact:version` I'd be happy to merge it.

**Comment by kshchepanovskyi** [comment]
It might be safer to apply something like https://maven.apache.org/plugins/maven-shade-plugin/

Anyway, I was thinking that version conflicts should not happen as all plugins are [loaded using separate classloaders](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206131249-XmlRpc-library-conflict):

> plugins are loaded in separate class loaders in a special manner to prevent conflicts between plugin and core versions of same libraries

Doesn't it work anymore?

**Comment by YannCebron** [comment]
related https://youtrack.jetbrains.com/issue/MP-2427

**Comment by YannCebron** [comment]
FTR the list of bundled 3rd party libraries is available from https://www.jetbrains.com/legal/third-party-software/

**Comment by hsz** [comment]
Mapping everything under `idea/lib` to Maven coordinates isn't reliable (no stable filename → group:artifact:version mapping), and bundled platform libraries typically load in separate classloaders, so a generic `failOnVersionConflict()` against the whole IDE lib set isn't something we'll do in this plugin.

What we *do* check today is the high-impact cases in `verifyPluginProjectConfiguration` (Kotlin stdlib, kotlinx-coroutines, Kotlin version vs. target platform). Bundled third-party libraries are also listed at https://www.jetbrains.com/legal/third-party-software/

If you're hitting a concrete conflict on a current 2.x version, please open a new issue with a repro.
